backends/tc/ebpfCodeGen.cpp

```diff
@@ -0,0 +1,1481 @@
+
+    builder->emitIndent();
+    builder->appendFormat("return %d;", actUnspecCode);
+    builder->newline();
+    builder->blockEnd(true);
+    builder->target->emitCodeSection(builder, sectionName);
+    builder->emitIndent();
+    builder->appendFormat("int %s(%s *%s)", functionName, builder->target->packetDescriptorType(),
+                          model.CPacketName.str());
+    builder->spc();
+
+    builder->blockStart();
+
+    emitGlobalMetadataInitializer(builder);
+
+    emitHeaderInstances(builder);
+    builder->newline();
+
+    builder->emitIndent();
+    builder->appendFormat("int ret = %d;", actUnspecCode);
+    builder->newline();
+    builder->emitIndent();
+    builder->appendFormat("int i;");
+    builder->newline();
+    builder->emitIndent();
+    builder->appendLine("#pragma clang loop unroll(disable)");
+    builder->emitIndent();
+    builder->appendFormat("for (i = 0; i < %d; i++) ", maxResubmitDepth);
+    builder->blockStart();
+    builder->emitIndent();
+    builder->append("ret = process(skb, ");
+
+    builder->appendFormat("(%s %s *) %s, %s);",
+                          parser->headerType->to<EBPF::EBPFStructType>()->kind,
+                          parser->headerType->to<EBPF::EBPFStructType>()->name,
+                          parser->headers->name.name, compilerGlobalMetadata);
+
+    builder->newline();
+    builder->appendFormat(
+        "        if (%s->drop == 1) {\n"
+        "            break;\n"
+        "        }\n",
+        compilerGlobalMetadata);
+    builder->blockEnd(true);
+    builder->newline();
+    builder->emitIndent();
+    builder->appendFormat("%s->recirculated = (i > 0);", compilerGlobalMetadata);
+    builder->newline();
+    builder->emitIndent();
+    builder->appendFormat(
+        "if (ret != %d) {\n"
+        "        return ret;\n"
+        "    }",
+        actUnspecCode);
+    builder->newline();
+
+    this->emitTrafficManager(builder);
+
+    builder->blockEnd(true);
+}
+
+void TCIngressPipelinePNA::emitGlobalMetadataInitializer(EBPF::CodeBuilder *builder) {
+    builder->emitIndent();
+    builder->appendFormat(
+        "struct pna_global_metadata *%s = (struct pna_global_metadata *) skb->cb;",
+        compilerGlobalMetadata);
+    builder->newline();
+
+    // if Traffic Manager decided to pass packet to the kernel stack earlier, send it up immediately
+    builder->emitIndent();
+    builder->append("if (compiler_meta__->pass_to_kernel == true) return TC_ACT_OK;");
+    builder->newline();
+
+    // workaround to make TC protocol-independent, DO NOT REMOVE
+    builder->emitIndent();
+    // replace ether_type only if a packet comes from XDP
+    builder->appendFormat("if (!%s->recirculated) ", compilerGlobalMetadata);
+    builder->blockStart();
+    builder->emitIndent();
+    builder->appendFormat("compiler_meta__->mark = %u", packetMark);
+    builder->endOfStatement(true);
+    builder->emitIndent();
+    // XDP2TC mode is assumed as META
```

Comment:
Some NIC drivers don't support the `bpf_xdp_adjust_meta()` helper (e.g., some of Mellanox drivers, AFAIR). That's why we created `head` and `cpumap` modes. I'd suggest keeping them.

Reply:
For now since, generateToXDP option is not supported, code for xdp2tcMode as XDP2TC_META is retained. When this option is added, Code for them will be added

Reply:
Updated the code